wire-grpc-client/src/main/java/com/squareup/wire/internal/BlockingMessageSource.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.internal
+
+import com.squareup.wire.MessageSource
+import com.squareup.wire.ProtoAdapter
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.Response
+import java.io.IOException
+import java.util.concurrent.LinkedBlockingDeque
+
+/**
+ * This message source uses a [LinkedBlockingDeque] to connect a reading source with a writing
+ * response callback.
+ *
+ * It uses two sentinel types:
+ *
+ *  * Failure: enqueued when there's an exception on the stream.
+ *  * Complete: enqueued when the stream completes normally.
+ */
+internal class BlockingMessageSource<R : Any>(
+  val call: Call,
+  val responseAdapter: ProtoAdapter<R>
+) : MessageSource<R> {
+  private val queue = LinkedBlockingDeque<Any>(1)
+
+  override fun read(): R? {
+    return when (val result = queue.take()) {
+      is Complete -> {
+        queue.put(result) // Replace it.
+        null
+      }
+      is Failure -> {
+        queue.put(result) // Replace it.
+        throw result.e
+      }
+      else -> result as R
+    }
+  }
+
+  override fun close() {
+    call.cancel() // Short-circuit the request stream if it's still active.
+  }
+
+  /** Read messages from the response body and write them to the deque. */
+  fun readFromResponseBodyCallback(): Callback {
+    return object : Callback {
+      override fun onFailure(call: Call, e: IOException) {
+        queue.put(e)
```

Comment:
```suggestion
        queue.put(Failure(e))
```

Reply:
oooh good catch. Test coverage is not there.